pageserver/src/bin/pageserver.rs

```diff
@@ -277,12 +277,30 @@ fn start_pageserver(conf: &'static PageServerConf) -> anyhow::Result<()> {
         }
     };
 
-    let remote_storage = conf
-        .remote_storage_config
-        .as_ref()
-        .map(GenericRemoteStorage::from_config)
-        .transpose()
-        .context("Failed to init generic remote storage")?;
+    let remote_storage = if let Some(remote_storage_conf) = &conf.remote_storage_config {
+        let remote_storage = GenericRemoteStorage::from_config(remote_storage_conf)
+            .context("Failed to init generic remote storage")?;
+
+        if conf.test_remote_failures > 0 {
+            if cfg!(feature = "testing") {
+                info!(
+                    "Simulating remote failures for first {} attempts of each op",
+                    conf.test_remote_failures
+                );
+                Some(GenericRemoteStorage::unreliable_wrapper(
+                    remote_storage,
+                    conf.test_remote_failures,
+                ))
+            } else {
+                anyhow::bail!("test_remote_failures option is not available because pageserver was compiled without the 'testing' feature");
```

Comment:
Is it possible to hide this test code in a RemoteStorageFactory that will be the only place to mix test and production code?

Reply:
Hmm, I guess it would make to put this in a separate function, the `start_pageserver` function is quite long already.

I also opened https://github.com/neondatabase/neon/pull/3145 for some minor cleanups around here.

Reply:
Moved this initialization to a separate function